File: cmd/genpack/schemajson.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
)

// This file turns `terraform providers schema -json` output into the
// attribute surface half of a pack.
//
// Why generate rather than curate: the hand-written schema this replaces
// listed 29 arguments for aws_instance; the provider actually declares 47
// plus 16 nested block types. Every argument missing from a curated list is
// a false "hallucinated attribute" finding on valid Terraform — and since
// that rule is severity high, a false positive there blocks a PR. Curation
// cannot keep up with a provider that ships every week, so it shouldn't try.

// tfProvidersSchema is the top level of `terraform providers schema -json`.
type tfProvidersSchema struct {
	FormatVersion   string                     `json:"format_version"`
	ProviderSchemas map[string]*tfProviderItem `json:"provider_schemas"`
}

type tfProviderItem struct {
	ResourceSchemas map[string]*tfResourceSchema `json:"resource_schemas"`
}

type tfResourceSchema struct {
	Version int      `json:"version"`
	Block   *tfBlock `json:"block"`
}

type tfBlock struct {
	Attributes map[string]*tfAttribute `json:"attributes"`
	BlockTypes map[string]*tfBlockType `json:"block_types"`
}

type tfAttribute struct {
	Required bool `json:"required"`
	Optional bool `json:"optional"`
	Computed bool `json:"computed"`
}

type tfBlockType struct {
	NestingMode string   `json:"nesting_mode"`
	Block       *tfBlock `json:"block"`
}

// metaArguments are valid inside any resource block but are Terraform's own,
// so they never appear in a provider's schema. Without them every `count`,
// `for_each` or `lifecycle` in a scanned repo would read as a hallucination.
var metaArguments = []string{
	"count",
	"depends_on",
	"for_each",
	"lifecycle",
	"provider",
	"provisioner",
	"connection",
	"dynamic",
}

// loadProviderSchema reads the terraform-generated schema JSON and returns
// the attribute surface per resource type, keyed the same way a Pack is.
func loadProviderSchema(path, providerAddr string) (map[string]*PackResource, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading provider schema: %w", err)
	}

	var doc tfProvidersSchema
	if err := json.Unmarshal(raw, &doc); err != nil {
		return nil, fmt.Errorf("parsing provider schema JSON: %w", err)
	}

	item := doc.ProviderSchemas[providerAddr]
	if item == nil {
		available := make([]string, 0, len(doc.ProviderSchemas))
		for k := range doc.ProviderSchemas {
			available = append(available, k)
		}
		sort.Strings(available)
		return nil, fmt.Errorf("provider %q not in schema (found: %s)", providerAddr, strings.Join(available, ", "))
	}

	out := make(map[string]*PackResource, len(item.ResourceSchemas))
	for rType, rSchema := range item.ResourceSchemas {
		if rSchema.Block == nil {
			continue
		}
		res := &PackResource{NestedBlocks: map[string][]string{}}
		collectBlock(rSchema.Block, "", res)

		// Meta-arguments only apply at the resource's top level.
		res.TopLevel = append(res.TopLevel, metaArguments...)
		res.TopLevel = dedupe(res.TopLevel)

		if len(res.NestedBlocks) == 0 {
			res.NestedBlocks = nil
		}
		out[rType] = res
	}
	return out, nil
}

// collectBlock walks a block and its nested blocks recursively, recording the
// valid argument names at each dotted path.
//
// A nested block's own name counts as a valid argument of its parent: HCL
// allows both `root_block_device { ... }` (block syntax) and, for some
// nesting modes, `root_block_device = [...]` (attribute syntax). Treating the
// name as valid in both positions avoids flagging a legal spelling.
func collectBlock(b *tfBlock, path string, res *PackResource) {
	names := make([]string, 0, len(b.Attributes)+len(b.BlockTypes))
	for name := range b.Attributes {
		names = append(names, name)
	}
	for name := range b.BlockTypes {
		names = append(names, name)
	}

	if path == "" {
		res.TopLevel = append(res.TopLevel, names...)
	} else {
		res.NestedBlocks[path] = names
	}

	for name, bt := range b.BlockTypes {
		if bt.Block == nil {
			continue
		}
		child := name
		if path != "" {
			child = path + "." + name
		}
		collectBlock(bt.Block, child, res)
	}
}

func dedupe(in []string) []string {
	seen := make(map[string]bool, len(in))
	out := in[:0]
	for _, v := range in {
		if seen[v] {
			continue
		}
		seen[v] = true
		out = append(out, v)
	}
	return out
}

File: src/tfpdf/genpack/schemajson.py
"""Transformation de la sortie de `terraform providers schema -json` en la
moitié « surface d'attributs » d'un pack.

Port de cmd/genpack/schemajson.go.

Pourquoi générer plutôt que curer : le schéma écrit à la main que ceci remplace
listait 29 arguments pour `aws_instance` ; le fournisseur en déclare en réalité
47, plus 16 types de blocs imbriqués. Chaque argument absent d'une liste curée
est une fausse découverte « attribut halluciné » sur du Terraform valide — et
comme cette règle est de sévérité haute, un faux positif y bloque une PR. La
curation ne peut pas suivre un fournisseur qui publie chaque semaine, elle ne
devrait donc pas essayer.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .pack import PackResource

#: Valid inside any resource block but Terraform's own, so they never appear in
#: a provider's schema. Without them every `count`, `for_each` or `lifecycle` in
#: a scanned repo would read as a hallucination.
META_ARGUMENTS = (
    "count",
    "depends_on",
    "for_each",
    "lifecycle",
    "provider",
    "provisioner",
    "connection",
    "dynamic",
)


class SchemaError(ValueError):
    """Le document de schéma du fournisseur n'a pas pu être lu, ou ne décrit pas
    le fournisseur demandé."""


def load_provider_schema(path: str | Path, provider_addr: str) -> dict[str, PackResource]:
    """Lit le JSON de schéma généré par terraform et rend la surface
    d'attributs par type de ressource, indexée comme l'est un pack."""
    try:
        raw = Path(path).read_bytes()
    except OSError as exc:
        raise SchemaError(f"reading provider schema: {exc}") from exc
    try:
        document = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise SchemaError(f"parsing provider schema JSON: {exc}") from exc
    if not isinstance(document, dict):
        raise SchemaError("parsing provider schema JSON: top level is not an object")

    schemas = document.get("provider_schemas") or {}
    item = schemas.get(provider_addr)
    if item is None:
        available = ", ".join(sorted(schemas))
        raise SchemaError(f'provider "{provider_addr}" not in schema (found: {available})')

    out: dict[str, PackResource] = {}
    for r_type, r_schema in (item.get("resource_schemas") or {}).items():
        block = (r_schema or {}).get("block")
        if not isinstance(block, dict):
            continue
        res = PackResource()
        _collect_block(block, "", res)

        # Meta-arguments only apply at the resource's top level.
        res.top_level = dedupe([*res.top_level, *META_ARGUMENTS])
        out[r_type] = res
    return out


def _collect_block(b: dict[str, Any], path: str, res: PackResource) -> None:
    """Parcourt récursivement un bloc et ses blocs imbriqués, en enregistrant
    les noms d'arguments valides à chaque chemin pointé.

    Le nom propre d'un bloc imbriqué compte comme un argument valide de son
    parent : HCL autorise à la fois `root_block_device { ... }` (syntaxe de bloc)
    et, pour certains modes d'imbrication, `root_block_device = [...]` (syntaxe
    d'attribut). Traiter le nom comme valide dans les deux positions évite de
    signaler une écriture légale.
    """
    attributes = b.get("attributes") or {}
    block_types = b.get("block_types") or {}
    names = [*attributes.keys(), *block_types.keys()]

    if path == "":
        res.top_level.extend(names)
    else:
        res.nested_blocks[path] = names

    for name, bt in block_types.items():
        child_block = (bt or {}).get("block")
        if not isinstance(child_block, dict):
            continue
        child = name if path == "" else path + "." + name
        _collect_block(child_block, child, res)


def dedupe(items: list[str]) -> list[str]:
    """Déduplication où la première occurrence l'emporte, en préservant
    l'ordre."""
    seen: set[str] = set()
    out: list[str] = []
    for v in items:
        if v not in seen:
            seen.add(v)
            out.append(v)
    return out
